contributors/design-proposals/node/node-usernamespace-remapping.md

```diff
@@ -0,0 +1,210 @@
+- The effort will be successful when it will be possible to use any container image, built using any set of uids and run it in Kubernetes cluster safely with host/node  uids not matching those in the image. Specifically, Pods with containers that need to be run as root (uid 0) should be runnable with non-root uids.
+- To allow images using privileged users securely where such containers will run in a user namespace different from user-namespace of the underlying host.  
+
+## Non-Goals
+- Non-goal is to support use cases when the uids for processes get allocated in runtime and externally (IPA, Active Directory) for generic users, without specifying (hopefully narrow) range of uids to be mapped upfront, before the container gets started.
+- Non-goal is to support pod/container level user namespace isolation. There can be images using different users but on the node, pods/containers running with these images will share common user namespace remapping configuration. In other words, all containers on a node share a common user-namespace range.
+- To support Pod-Security-Policy integration is a non-goal for this KEP.
+- To add any UX for allowing a mechanism where user namespace or host namespace can be requested in the pod spec, is a non-goal.
+
+## User Stories
+- As a cluster admin, I want to protect the node from the rogue container process(es) running inside pod containers with root privileges. If such a process is able to break out into the node, it could be a security issue.
+- As a cluster admin, I want to support all the images irrespective of what user/group that image is using.
+- As a cluster admin, I want to allow some pods to disable user namespaces if they require elevated privileges.
+
+## Proposal
+Proposal is to support user-namespaces for the pod containers. This can be done at two levels:
+- Node-level : This proposal explains this part in detail.
+- Pod-level: Plan is to target this in future due to missing support in the low level system components such as runtimes and kernel. More on this in the later sections.
+
+Node-level  user-namespace support means that, if feature is enabled, all pods on a node will share a common user-namespace, common UID(and GID) range (which is a subset of node’s total UIDs(and GIDs)). This common user-namespace is runtime’s default user-namespace range which is remapped to containers’ UIDs(and GID), starting with the first UID as container’s ‘root’.
+
+User-namespace support should be enabled only when container runtime on the node supports user-namespace remapping and is enabled in its configuration. To enable user-namespaces, feature-gate flag will need to be passed to Kubelet like this `--feature-gates=”NodeUserNamespace=true”`
+
+A new CRI API, `RuntimeConfigInfo` will be added. Kubelet will use this API:
```

Comment:
This should be Linux-specific (or have a linux-specific sub object)

Reply:
:+1: We can update it.

Reply:
@tallclair, would you recommend renaming `UserNamespaceConfig` to `LinuxUserNamespaceConfig` and potentially putting it directly to `RuntimeConfigInfoResponse`?

Reply:
Ah, now I see you propose `LinuxRuntimeConfig` below.